### PR TITLE
[Bug Fix] Fix tensor dump not saving when model uses .forward() calls

### LINKED ISSUE #26269: [Bug] Tensor dump produces no .pt files for models using .forward() internally
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Describe the bug

  When using `--debug-tensor-dump-output-folder`, the dump directory structure is created correctly but **no `.pt` files are generated** for models that call submodules via `.forward()` instead of `__call__()`. Inference itself works fine — only the dump is silently broken.
  Affected models include DeepSeek-V4 and any model where the top-level `forward()` calls inner modules with `.forward()`. Models like Llama that use `self.model(...)` internally are unaffected.

## Steps to reproduce

> For Docker image and launch reference, see: https://docs.sglang.io/cookbook/autoregressive/DeepSeek/DeepSeek-V4

  ```bash
  export CUDA_VISIBLE_DEVICES=0,1

  sglang serve \
    --trust-remote-code \
    --model-path deepseek-ai/DeepSeek-V4-Flash \
    --tp 2 \
    --enable-dp-attention \
    --disable-cuda-graph \
    --debug-tensor-dump-output-folder /workspace/dump_sglang/ \
    --port 7000
  ```

  Send a request:

  ```bash
  curl http://127.0.0.1:7000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "deepseek-ai/DeepSeek-V4-Flash",
      "temperature": 0,
      "seed": 42,
      "messages": [{"role": "user", "content": "What is 15% of 240?"}]
    }'
  ```

  Inference response:

  ```json
  {"id":"14660471ddf94ec489addbc44d599186","object":"chat.completion","created":1778398919,"model":"deepseek-ai/DeepSeek-V4-Flash","choices":[{"index":0,"message":{"role":"assistant","content":"To find 15% of 240, you can multiply 240 by 0.15 (since 15% = 15/100 = 0.15).\n\n\\[\n240 \\times 0.15 = 36\n\\]\n\nAlternatively, you can find 10% of 240 (which is 24) and 5% of 240 (which is 12), then add them together:  \n24 + 12 = 36.\n\nSo, 15% of 240 is **36**.","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":1}],"usage":{"prompt_tokens":13,"total_tokens":119,"completion_tokens":106,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
```

  The dump directory is created but contains no `.pt` files:

Image

  Result: Response is correct, but /workspace/dump_sglang/TP0_PP0_Rank0_pidXXX/ contains no .pt files.
  
  Expected: Each forward pass should produce a PassXXXXX.pt file.

Root cause analysis

  1. PyTorch hook trigger condition

  register_forward_hook only fires when a module is invoked via __call__():

```bash
  module(input)          # ✅ triggers hooks (__call__ → pre_hook → forward → post_hook)
  module.forward(input)  # ❌ bypasses hooks (calls forward directly)
```

  2. The call chain in SGLang
  
```bash
  model_runner.py:
    self.model.forward(...)              ← bypasses __call__, outermost hook never fires
      → ForCausalLM.forward():
          self.model.forward(...)        ← bypasses __call__, inner top-level hook never fires
            → Model.forward():
                → leaf modules via __call__()  ← hooks fire, tensors collected ✅
```

  The tensor dump system has two roles:
  - Leaf hooks (do_dump=False): collect tensor outputs into _current_tensors — these work fine
  - Top-level hook (do_dump=True): calls dump_current_tensors() to save collected data as .pt — this never fires

  So data is collected but never written to disk.

  3. Verification
  
To pinpoint where the pipeline breaks, I added logging at three points in `tensor_dump_forward_hook.py`:

  **Step 1: Check if leaf-node hooks fire** — Added a print inside `inner_dump_hook` when `do_dump=False`.

  Result: ✅ All leaf hooks fire as expected, `add_tensor()` is called for every submodule output.

Image

**Step 2: Check if `_current_tensors` is populated** — Printed `len(self._current_tensors)` after each `add_tensor()` call.

  Result: ✅ Data accumulates correctly (50+ tensors per forward pass).

Image

**Step 3: Check if the top-level hook fires** — Added a print inside `inner_dump_hook` when `do_dump=True`.

  Result: ❌ This print never appears. The hook registered on `DeepseekV4Model` is never triggered because it's called via `.forward()`, not
  `__call__()`. As a result, `dump_current_tensors()` is never executed.

Image

  **Conclusion**: The pipeline breaks between collection and writing — tensors are gathered but the flush-to-disk step is never reached.

  ## Fix

  I have a fix ready and will submit a PR shortly.

## Steps to reproduce

> For Docker image and launch reference, see: https://docs.sglang.io/cookbook/autoregressive/DeepSeek/DeepSeek-V4

  ```bash
  export CUDA_VISIBLE_DEVICES=0,1

  sglang serve \
    --trust-remote-code \
    --model-path deepseek-ai/DeepSeek-V4-Flash \
    --tp 2 \
    --enable-dp-attention \
    --disable-cuda-graph \
    --debug-tensor-dump-output-folder /workspace/dump_sglang/ \
    --port 7000
  ```

  Send a request:

  ```bash
  curl http://127.0.0.1:7000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "deepseek-ai/DeepSeek-V4-Flash",
      "temperature": 0,
      "seed": 42,
      "messages": [{"role": "user", "content": "What is 15% of 240?"}]
    }'
  ```

  Inference response:

  ```json
  {"id":"14660471ddf94ec489addbc44d599186","object":"chat.completion","created":1778398919,"model":"deepseek-ai/DeepSeek-V4-Flash","choices":[{"index":0,"message":{"role":"assistant","content":"To find 15% of 240, you can multiply 240 by 0.15 (since 15% = 15/100 = 0.15).\n\n\\[\n240 \\times 0.15 = 36\n\\]\n\nAlternatively, you can find 10% of 240 (which is 24) and 5% of 240 (which is 12), then add them together:  \n24 + 12 = 36.\n\nSo, 15% of 240 is **36**.","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":1}],"usage":{"prompt_tokens":13,"total_tokens":119,"completion_tokens":106,"prompt_tokens_details":null,"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
```

### Environment

Environment

  - SGLang: latest main
  - Model: DeepSeek-V4-Flash (any model using .forward() internally is affected)
  - PyTorch: 2.x
  - GPU: NVIDIA B200

### PR DESCRIPTION
**Pull request by yaya159456**
Fixes #26269

Fix by:
1. Moving the dump trigger from the inner top-level module hook to a new root hook registered on the outermost model (guaranteed to fire since model_runner now uses `__call__()`)
2. Changing all `self.model.forward(...)` calls in model_runner to `self.model(...)` so the root hook triggers via `__call__()`

## Motivation

The `--debug-tensor-dump-output-folder` feature silently fails to produce any `.pt` files for models that internally call submodules via `.forward()` (e.g., DeepSeek-V4). The dump directory structure is created but remains empty.
  **Root cause**: PyTorch's `register_forward_hook` only fires when a module is invoked via `__call__()`. In `model_runner.py`, the outermost model is called with `self.model.forward(...)`, which bypasses `__call__()` and prevents the top-level dump hook from triggering. The leaf-node hooks collect tensors into `_current_tensors`, but `dump_current_tensors()` is never called — data is gathered but never written to disk.

  Models like Llama that use `self.model(...)` internally are unaffected since the hook fires normally.

## Modifications

**`python/sglang/srt/debug_utils/tensor_dump_forward_hook.py`**:
  - Remove `self.dump_current_tensors()` from the inner `_dump_hook` (it was unreachable for `.forward()`-style models anyway)
  - Add a `_root_dump_hook` registered on the outermost model passed to `register_forward_hook_for_model()`, which calls `dump_current_tensors()` after all sub-hooks have collected their data

  **`python/sglang/srt/model_executor/model_runner.py`**:
  - Change 4 occurrences of `self.model.forward(...)` to `self.model(...)` so the outermost model is invoked via `__call__()`, ensuring the root hook fires

## Verification

  After the fix, tensor dump works correctly for DeepSeek-V4-Flash:
  **TP0 (Rank0)**:
  
image

  **TP1 (Rank1)**:
 
image

 **Before vs After**:

  | Metric | Before | After |
  |--------|--------|-------|
  | Directory creation | ✅ Normal | ✅ Normal |
  | `.pt` file generation | ❌ Empty directory | ✅ One file per forward pass |
  | Multi-TP rank | ❌ No files on any rank | ✅ Each rank generates complete files independently |

  **Data completeness per `.pt` file**:

  | Model type | Before | After |
  |------------|--------|-------|
  | Llama (uses `__call__()`) | ✅ Works, but `logits_processor` delayed to next file | ✅ Works, all data in same file |
  | DeepSeek V4 (uses `.forward()`) | ❌ No `.pt` files at all | ✅ All data in same file |

  - `.pt` files generated as expected (Pass00000.pt, Pass00001.pt, ...)
  - All tensor data complete and saved in correct per-pass files

  Models using `__call__()` internally (e.g., Llama) continue to work identically — the root hook simply takes over the dump responsibility from the inner hook, with no change in behavior or output.

## Accuracy Tests

This change does not affect model outputs — it only fixes the tensor dump I/O path. The model forward computation is identical.

## Speed Tests and Profiling

No performance impact — the only added overhead is one additional `register_forward_hook` call on the outermost model, which is negligible.

## Checklist

- [x] Format: `pre-commit run --all-files`
- [x] The fix is model-agnostic — no changes to any `models/*.py` file
- [x] Backward compatible: models that already worked (e.g., Llama) continue to work identically
- [x] Bonus: `logits_processor` output is now saved in the same `.pt` file as the rest of the forward pass (previously it was delayed to the
  next file)

 ## Additional context

  Reproducible with any model that uses `.forward()` internally (confirmed on DeepSeek-V4-Flash with `--disable-cuda-graph`). The issue is
  hidden when CUDA Graph is enabled because the graph replay path doesn't go through the Python forward hooks at all.

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #26390502994](https://github.com/sgl-project/sglang/actions/runs/26390502994)
Latest PR Test (Extra): :x: [Run #26390502878](https://github.com/sgl-project/sglang/actions/runs/26390502878)

### CODE REVIEW

File: python/sglang/srt/debug_utils/tensor_dump_forward_hook.py [hunk 1]
```diff
@@ -161,4 +160,9 @@ def register_forward_hook_for_model(
     assert (
         model_top_level_module_matched
     ), f"model should have a module named {top_level_module_name}"
+
+    def _root_dump_hook(module, input, output):
+        tensor_dumper.dump_current_tensors()
+
+    model.register_forward_hook(_root_dump_hook)
```
[the remark this answers is not in the document]
Comment (yaya159456) [reply python/sglang/srt/debug_utils/tensor_dump_forward_hook.py:167]: Thanks for the suggestion. In practice, if `forward()` raises an exception in SGLang's model_runner, the request fails entirely and the worker
   doesn't continue processing subsequent requests with the same state. The stale tensor scenario is unlikely to occur in real usage.

  I'd prefer to keep this PR focused on the core fix. If this edge case becomes a concern, it can be addressed in a follow-up.

[the remark this answers is not in the document]
Comment (yaya159456) [reply python/sglang/srt/debug_utils/tensor_dump_forward_hook.py:167]: Also, the original code didn't have this protection either — this is a pre-existing edge case, not a regression introduced by this PR.
